source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opt/scalar_analysis.h"
+#include <algorithm>
+#include <functional>
+#include <string>
+#include <utility>
+
+#include "opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+SENode* ScalarEvolutionAnalysis::CreateNegation(SENode* operand) {
+  if (operand->GetType() == SENode::Constant) {
+    return CreateConstant(-operand->AsSEConstantNode()->FoldToSingleValue());
+  }
+  std::unique_ptr<SENode> negation_node{new SENegative(this)};
+  negation_node->AddChild(operand);
+  return GetCachedOrAdd(std::move(negation_node));
+}
+
+SENode* ScalarEvolutionAnalysis::CreateConstant(int64_t integer) {
+  return GetCachedOrAdd(
+      std::unique_ptr<SENode>(new SEConstantNode(this, integer)));
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzeMultiplyOp(
+    const ir::Instruction* multiply) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+
+  SENode* op1 =
+      AnalyzeInstruction(def_use->GetDef(multiply->GetSingleWordInOperand(0)));
+  SENode* op2 =
+      AnalyzeInstruction(def_use->GetDef(multiply->GetSingleWordInOperand(1)));
+
+  return CreateMultiplyNode(op1, op2);
+}
+
+SENode* ScalarEvolutionAnalysis::CreateMultiplyNode(SENode* operand_1,
+                                                    SENode* operand_2) {
+  if (operand_1->GetType() == SENode::Constant &&
+      operand_2->GetType() == SENode::Constant) {
+    return CreateConstant(operand_1->AsSEConstantNode()->FoldToSingleValue() *
+                          operand_2->AsSEConstantNode()->FoldToSingleValue());
+  }
```

Comment:
I'm wondering whether it wouldn't make better sense to have a helper function that simplifies SENode expression trees in a similar way as the instruction folder.  It does not need to handle many cases, but it may simplify the building of the SENode expressions.

Reply:
We do that in the SimplifyExpression method. It's implemented in scalar_analysis_simplification.cpp. The reason we sill do some of these optimisations on creation is that they're fairly cheap to do and it simplifies some of the code later on (for instance knowing that a negate will never have a constant as child)